feat(elasticache): remove userName property from IamUser construct

sAchin-680 opened the pull request:
feat(elasticache): remove userName property from IamUser construct

## Issue # (if applicable)

Closes #35848

## Reason for this change

For IAM-enabled ElastiCache users, the userName and userId properties must be identical as per AWS requirements. The current implementation allows users to specify userName in the constructor props, but it must always equal userId for IAM authentication. This creates unnecessary complexity and potential for configuration errors, as users might mistakenly provide different values.

## Description of changes

Fixed the API design by removing the userName property from IamUserProps interface.
- Removed userName from IamUserProps interface to prevent configuration errors
- Modified IamUser constructor to automatically set userName = userId internally
- Updated comprehensive test suite with integration test exemption documentation
- Maintained backward compatibility for import methods (fromUserAttributes, etc.) since they represent external resources

## Describe any new or updated permissions being added

No new or updated IAM permissions are being added with these changes.

## Description of how you validated changes

- Verified that all existing unit tests pass after the changes
- Updated test cases to remove userName property from IamUser constructor calls
- Added comprehensive test coverage for automatic userName assignment behavior
- Confirmed CloudFormation template generation produces correct UserName property values
- Added integration test exemption documentation with detailed reasoning

## Checklist

- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

**BREAKING CHANGE**: The userName property in IamUserProps has been removed. Users should remove userName from IamUser constructor calls - it will be automatically set to userId value.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
The pull request linter fails with the following errors:

    ❌ Features must contain a change to an integration test file and the resulting snapshot.

If you believe this pull request should receive an exemption, please comment and provide a justification. A comment requesting an exemption should contain the text `Exemption Request`. Additionally, if clarification is needed, add `Clarification Request` to a comment.

mazyu36 reviewed (commented):
Thank you for your contribution.

For IamUser, it would be better to write in the README that userName is automatically set.

mazyu36 commented on packages/@aws-cdk/aws-elasticache-alpha/test/iam-user.test.ts:
```diff
@@ -52,12 +30,32 @@ describe('IamUser', () => {
+    test('automatically sets userName to userId regardless of provided userName', () => {
+      const user = new IamUser(stack, 'TestUser', {
+        userId: 'test-user-id',
+        userName: 'different-name', // This should be ignored
```
userName in unit test should be removed.
```suggestion
```

sAchin-680 replied:
I've removed the `userName` from the unit test as suggested.

mazyu36 commented:
Also, please add a comment with an Exemption Request regarding not adding integration tests.
I think it's fine without them, but the maintainers might have a different opinion.

sAchin-680 commented:
> Thank you for the revisions. I've added a few additional comments.
> 
> Also, since this is a breaking change, could you please add a note about the breaking change to the PR description, similar to the following PR?
> 
> #35920

Thank you for the feedback and for pointing that out !! 
I’ve added the breaking change note to the PR description, following the example from #35920.

sAchin-680 commented on packages/@aws-cdk/aws-elasticache-alpha/lib/iam-user.ts:
```diff
@@ -3,22 +3,21 @@ import { UserEngine } from './common';
-   * @default - Same as userId.
-   */
-  readonly userName?: string;
+  // userName property removed - it's automatically set to userId internally
```
You're right @mazyu36, that comment isn’t needed. I've removed the line.

mazyu36 replied:
It looks like it hasn't been removed yet.

sAchin-680 replied:
ok removed !

sAchin-680 commented:
Hey @mazyu36, @pahud, @aws-cdk-automation and team, just following up on this PR since it’s been pending for quite some time now. I’ve addressed the earlier feedback and updated everything from my side.

Would appreciate any further review/comments whenever you get a chance. Thanks!

mazyu36 commented:
@sAchin-680 
Hi.

From here, we'll need to wait for a maintainer to review. I'm a community reviewer, not a maintainer myself.

Also, it looks like the CI is currently failing, so it would be great if you could fix that first.

After that, could you wait for a maintainer review?

sAchin-680 commented:
thanks for the clarification @mazyu36 ! I’ve fixed the CI now, will wait for a maintainer review.
